src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,3315 @@
+    let _num_headers = headers.len();
+
+    let mut ret = vec![];
+    for header in headers.into_iter() {
+        let blk = BurnchainDB::get_burnchain_block(&burnchain_tx.conn(), &header.block_hash)
+            .expect(&format!(
+                "BUG: failed to load prepare-phase block {} ({})",
+                &header.block_hash, header.block_height
+            ));
+
+        let mut block_ops = vec![];
+        for op in blk.ops.into_iter() {
+            assert!(pox_consts.is_in_prepare_phase(first_block_height, op.block_height()));
+            match op {
+                BlockstackOperationType::LeaderBlockCommit(opdata) => {
+                    // basic validity filtering
+                    if opdata.block_height < first_block_height {
+                        test_debug!("Skip too-early block commit");
+                        continue;
+                    }
+                    if (opdata.parent_block_ptr as u64) < first_block_height {
+                        if opdata.parent_block_ptr != 0 || opdata.parent_vtxindex != 0 {
+                            test_debug!("Skip orphaned block-commit");
+                            continue;
+                        }
+                    }
+                    if opdata.block_height <= opdata.parent_block_ptr as u64 {
+                        test_debug!("Skip block-commit whose 'parent' comes at or after it");
+                        continue;
+                    }
+                    if opdata.burn_fee == 0 {
+                        test_debug!("Skip block-commit without burn");
+                        continue;
+                    }
+                    block_ops.push(opdata);
+                }
+                _ => {
+                    continue;
+                }
+            }
+        }
+        block_ops.sort_by(|op1, op2| {
+            if op1.block_height != op2.block_height {
+                op1.block_height.cmp(&op2.block_height)
+            } else {
+                op1.vtxindex.cmp(&op2.vtxindex)
+            }
+        });
+        ret.push(block_ops);
+    }
+
+    test_debug!(
+        "Read {} headers, {} prepare-phase commits from reward cycle {} ({}-{})",
+        _num_headers,
+        ret.len(),
+        reward_cycle,
+        start_height,
+        end_height
+    );
+    Ok(ret)
+}
+
+/// Find all referenced parent block-commits already in the burnchain DB, so we can extract their VRF seeds.
+pub fn read_parent_block_commits<'a, B: BurnchainHeaderReader>(
+    burnchain_tx: &BurnchainDBTransaction<'a>,
+    indexer: &B,
+    prepare_phase_ops: &Vec<Vec<LeaderBlockCommitOp>>,
+) -> Result<Vec<LeaderBlockCommitOp>, Error> {
+    let mut parents = HashMap::new();
+    for ops in prepare_phase_ops.iter() {
+        for opdata in ops.iter() {
+            let mut hdrs = indexer.read_burnchain_headers(
+                opdata.parent_block_ptr as u64,
+                (opdata.parent_block_ptr + 1) as u64,
+            )?;
```

Comment:
This method gets called a number of times to fetch a single burnchain header. It probably makes sense to just add another method `read_burnchain_header(height)`.

Reply:
Done